src/plugins/ui_actions/public/actions/action.ts

```diff
@@ -17,49 +17,132 @@
+  isCompatible(context: Context): Promise<boolean>;
 
   /**
    * If this returns something truthy, this is used in addition to the `execute` method when clicked.
    */
-  getHref?(context: ActionContext): string | undefined;
+  getHref?(context: Context): string | undefined;
+
+  /**
+   * Executes the action.
+   */
+  execute(context: Context): Promise<void>;
+}
+
+/**
+ * A convenience interface used to register an action.
+ */
+export interface ActionDefinition<
+  PresentationContext extends object = object,
+  ExecutionContext extends object = PresentationContext,
+  Return = Promise<void>,
+  Config extends object | undefined = undefined
+>
+  extends Partial<Presentable<PresentationContext>>,
+    Partial<Configurable<Config, PresentationContext>> {
+  /**
+   * ID of the action that uniquely identifies this action in the actions registry.
+   */
+  readonly id: string;
+
+  readonly type?: string;
+
+  /**
+   * ID of the `FactoryAction` that can be used to construct instances of
+   * this dynamic action.
+   */
+  readonly factoryId?: string;
+
+  /**
+   * @deprecated
+   *
+   * Do not use this, use `execute` method instead.
+   */
+  getHref?(context: ExecutionContext): string | undefined;
 
   /**
    * Executes the action.
    */
-  execute(context: ActionContext): Promise<void>;
+  execute(
+    context: ExecutionContext,
+    action: ActionContract<ActionDefinition<PresentationContext, ExecutionContext, Return, Config>>
+  ): Return;
+}
+
+export type AnyActionDefinition = ActionDefinition<any, any, any, any>;
+export type ActionPresentationContext<A> = A extends ActionDefinition<infer Context, any, any, any>
+  ? Context
+  : never;
+export type ActionExecutionContext<A> = A extends ActionDefinition<any, infer Context, any, any>
+  ? Context
+  : never;
+export type ActionConfig<A> = A extends ActionDefinition<any, any, infer Config> ? Config : never;
+
+/**
+ * A convenience interface used to register a dynamic action.
+ *
+ * A dynamic action is one that can be create by user and registered into the
+ * actions registry at runtime. User can also provide custom config for this
+ * action. And dynamic actions can be serialized for storage and deserialized
+ * back.
+ */
+export type DynamicActionDefinition<
+  PresentationContext extends object = object,
+  ExecutionContext extends object = PresentationContext,
+  Return = Promise<void>,
+  Config extends object | undefined = undefined
+> = ActionDefinition<PresentationContext, ExecutionContext, Return, Config> &
+  Required<
+    Pick<
+      ActionDefinition<PresentationContext, ExecutionContext, Return, Config>,
+      'CollectConfig' | 'defaultConfig' | 'factoryId'
+    >
+  >;
+
+export type AnyDynamicActionDefinition = DynamicActionDefinition<any, any, any, any>;
+
+/**
+ * Factory actions are actions used to create dynamic actions - their `execute`
+ * method returns a dynamic
+ */
+export interface FactoryActionDefinition<
```

Comment:
I don't think this is being used anywhere, that I see.  

I think an example plugin highlighting the use-age would be a little more helpful to me than the storybook (though that is also nice).  If you had that I think you would actually be using this by registering a factory action?   Maybe it's not necessary?

Reply:
Agreed to remove this from this PR, with a follow up work that implements action factories, possibly creating a new action factory registry.